embassy-nrf/src/saadc.rs

```diff
@@ -403,6 +427,56 @@ impl<'d, const N: usize> Saadc<'d, N> {
             Poll::Pending
         })
         .await;
+
+        on_drop.defuse();
+        Self::stop_sampling().await;
+    }
+
+    // Stop sampling and wait for it to stop in a blocking fashion
+    fn stop_sampling_immediately() {
+        let r = Self::regs();
+
+        compiler_fence(Ordering::SeqCst);
+
+        r.events_stopped.reset();
+        r.tasks_stop.write(|w| unsafe { w.bits(1) });
+
+        while r.events_stopped.read().bits() == 0 {}
+        r.events_stopped.reset();
+    }
+
+    // Stop sampling and wait for it to stop in a non-blocking fashino
+    async fn stop_sampling() {
```

Comment:
how long does stopping take? if it's fast (<100 clock cycles) then it's not worth doing it async because an asyn context switch is longer than that. Doing it blocking would be faster and smaller code.

Reply:
I've measured it at well under 45 cycles, probably 22, as measured by DWT::cycle_count() and comparing a code path that stops and disables but doesn't wait for EVENTS_STOPPED (45 cycles) with one that does (23 cycles). Give or take a few cycles.  This isn't documented, but I assume at least in the case of `sample()` where it would stop only following the completion of the requested number of samples, that it won't ever take longer.  Stopping while it's actively converting (i.e. interrupting a requested sequence) would likely take longer (possibly requiring waiting for at least the current conversion to complete).

Reply:
Sounds like we should just stop and block wait then. I’ll do the mod. It’ll make things even simpler!

Reply:
Commit https://github.com/embassy-rs/embassy/pull/928/commits/30641d05640a7a3cd4d71c70c09694a3eef825e0 takes care of always stopping the sampling whether the futures are canceled or not.